Web Inspector: Show gap overlays between items in masonry axis for grid-lanes layouts

stwrt opened the pull request:
#### ee295e3dba5eb4080dd80a45fdeeeef940613aee
<pre>
Web Inspector: Show gap overlays between items in masonry axis for grid-lanes layouts
<a href="https://bugs.webkit.org/show_bug.cgi?id=304579">https://bugs.webkit.org/show_bug.cgi?id=304579</a>
<a href="https://rdar.apple.com/problem/166984079">rdar://problem/166984079</a>

Reviewed by Devin Rousso.

For grid-lanes (masonry) layouts, the grid overlay now shows gaps between
items in the masonry axis direction. Previously, only the bounding lines
of the masonry axis were drawn, with no gap visualization between items.

The implementation collects all grid items and their bounding boxes, groups
them by their grid axis track, sorts them by position in the masonry axis,
and draws gap regions between consecutive items within each track.

* Source/WebCore/inspector/InspectorOverlay.cpp:
(WebCore::InspectorOverlay::buildGridOverlay):

Canonical link: <a href="https://commits.webkit.org/304861@main">https://commits.webkit.org/304861@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/668c3d39841778a4f248559def4db207aac98bb2

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/136696 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/159/builds/9055 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/47983 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/144424 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/89669 "Built successfully") | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🛠 ios-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/5aa3c6b8-123e-4cc1-8148-d68333fb34bd) 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/138568 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/155/builds/9756 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/156/builds/8901 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/104535 "Passed tests") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/60/builds/76328 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🛠 mac-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/9401d229-eed1-4b5f-ac3d-2bada7e14d9f) 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/139641 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/162/builds/7131 "Passed tests") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/18/builds/122487 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/85374 "Passed tests") | | [✅ 🛠 vision-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/749ecaf1-0313-4afc-bf63-72a950684c99) 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/154/builds/6775 "Passed tests") | [✅ 🧪 api-mac-debug](https://ews-build.webkit.org/#/builders/165/builds/4465 "Passed tests") | [✅ 🛠 wpe-cairo-libwebrtc](https://ews-build.webkit.org/#/builders/166/builds/5016 "Built successfully") | | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/116094 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/40675 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/147181 "Built successfully") | | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/153/builds/8739 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/41248 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/112889 "Passed tests") | | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/160/builds/8757 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/161/builds/7356 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/113218 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/28757 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/164/builds/6697 "Passed tests") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/118778 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/62875 "Built successfully") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/158/builds/8787 "Built successfully") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/36832 "Passed tests") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/157/builds/8508 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/72353 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/163/builds/8727 "Built successfully") | | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/152/builds/8579 "Built successfully") | | | | 
<!--EWS-Status-Bubble-End-->

dcrousso commented on Source/WebCore/inspector/InspectorOverlay.cpp:
```diff
@@ -1864,6 +1864,130 @@ std::optional<InspectorOverlay::Highlight::GridHighlightOverlay> InspectorOverla
+            auto margins = gridItem->marginBox();
+            absoluteRect.expand(FloatBoxExtent { margins.top(), margins.right(), margins.bottom(), margins.left() });
+
+            // Transform to root coordinates for proper overlay positioning.
```
most of these comments are redundant if you just read the next line of code (specifically either the variable name or what it's iterating over)

generally speaking we only like to add comments for things that are not obvious so id suggest removing most of these comments

dcrousso commented on Source/WebCore/inspector/InspectorOverlay.cpp:
```diff
@@ -1864,6 +1864,130 @@ std::optional<InspectorOverlay::Highlight::GridHighlightOverlay> InspectorOverla
+            // Transform to root coordinates for proper overlay positioning.
+            auto minCorner = localPointToRootPoint(containingView, absoluteRect.minXMinYCorner());
+            auto maxCorner = localPointToRootPoint(containingView, absoluteRect.maxXMaxYCorner());
+            FloatRect rootRect { minCorner, FloatSize { maxCorner.x() - minCorner.x(), maxCorner.y() - minCorner.y() } };
```
i think we can just do `maxCorner - minCorner` as there's a `operator-` for `FloatPoint`

dcrousso commented on Source/WebCore/inspector/InspectorOverlay.cpp:
```diff
@@ -1864,6 +1864,130 @@ std::optional<InspectorOverlay::Highlight::GridHighlightOverlay> InspectorOverla
+            auto gridArea = renderGrid.currentGrid().gridItemArea(*gridItem);
+            auto absoluteRect = FloatRect { gridItem->absoluteBoundingBoxRect(true) };
+            auto margins = gridItem->marginBox();
+            absoluteRect.expand(FloatBoxExtent { margins.top(), margins.right(), margins.bottom(), margins.left() });
```
NIT: i wanna say that `LayoutBoxExtent` (a.k.a. `RectEdges<LayoutUnit>`) can automatically convert to `FloatBoxExtent` (a.k.a. `RectEdges<float>`) because `LayoutUnit` has an `operator float` so i think we could just do
```cpp
absoluteRect.expand(gridItem->marginBox());
```

dcrousso commented on Source/WebCore/inspector/InspectorOverlay.cpp:
```diff
@@ -1864,6 +1864,130 @@ std::optional<InspectorOverlay::Highlight::GridHighlightOverlay> InspectorOverla
+            // Draw gaps between consecutive items in this track.
+            for (size_t i = 1; i < itemsInTrack.size(); ++i) {
+                auto* prevItem = itemsInTrack[i - 1];
+                auto* currItem = itemsInTrack[i];
```
Style: we dont normally like to abbreviate names like this

dcrousso replied:
since we never check for `nullptr` anywhere, why not dereference them here and avoid that repeated work?
```cpp
auto& prevItem = *itemsInTrack[i - 1];
auto& currItem = *itemsInTrack[i];
```

dcrousso commented on Source/WebCore/inspector/InspectorOverlay.cpp:
```diff
@@ -1864,6 +1864,130 @@ std::optional<InspectorOverlay::Highlight::GridHighlightOverlay> InspectorOverla
+                        { gapLeft, gapTop },
+                        { gapRight, gapTop },
+                        { gapRight, gapBottom },
+                        { gapLeft, gapBottom }
```
Style: missing `,`

dcrousso commented on Source/WebCore/inspector/InspectorOverlay.cpp:
```diff
@@ -1864,6 +1864,130 @@ std::optional<InspectorOverlay::Highlight::GridHighlightOverlay> InspectorOverla
+                    if (gapRight <= gapLeft)
+                        continue; // No horizontal overlap.
+
+                    gapQuad = FloatQuad {
```
NIT: I dont think the `FloatQuad` is necessary here since you've already given that type for `gapQuad`

dcrousso commented on Source/WebCore/inspector/InspectorOverlay.cpp:
```diff
@@ -1864,6 +1864,130 @@ std::optional<InspectorOverlay::Highlight::GridHighlightOverlay> InspectorOverla
+                    if (gapBottom <= gapTop)
+                        continue; // No vertical overlap.
+
+                    gapQuad = FloatQuad {
```
ditto

dcrousso commented on Source/WebCore/inspector/InspectorOverlay.cpp:
```diff
@@ -1864,6 +1864,130 @@ std::optional<InspectorOverlay::Highlight::GridHighlightOverlay> InspectorOverla
+                        { gapLeft, gapTop },
+                        { gapRight, gapTop },
+                        { gapRight, gapBottom },
+                        { gapLeft, gapBottom }
```
ditto

dcrousso commented on Source/WebCore/inspector/InspectorOverlay.cpp:
```diff
@@ -1864,6 +1864,130 @@ std::optional<InspectorOverlay::Highlight::GridHighlightOverlay> InspectorOverla
+        // For areMasonryRows(), items flow vertically (rows are masonry), so group by column track.
+        // For areMasonryColumns(), items flow horizontally (columns are masonry), so group by row track.
+        struct ItemInfo {
+            RenderBox* item;
```
i think this might need to be `CheckedRef<RenderBox>`?

dcrousso reviewed (approved):
nice work!

stwrt committed:
Web Inspector: Show gap overlays between items in masonry axis for grid-lanes layouts
https://bugs.webkit.org/show_bug.cgi?id=304579
rdar://problem/166984079

Reviewed by Devin Rousso.

For grid-lanes (masonry) layouts, the grid overlay now shows gaps between
items in the masonry axis direction. Previously, only the bounding lines
of the masonry axis were drawn, with no gap visualization between items.

The implementation collects all grid items and their bounding boxes, groups
them by their grid axis track, sorts them by position in the masonry axis,
and draws gap regions between consecutive items within each track.

* Source/WebCore/inspector/InspectorOverlay.cpp:
(WebCore::InspectorOverlay::buildGridOverlay):

Canonical link: https://commits.webkit.org/304861@main
